core/src/main/java/org/apache/iceberg/taskio/BaseTaskWriter.java

```diff
@@ -0,0 +1,141 @@
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.Tasks;
+
+abstract class BaseTaskWriter<T> implements TaskWriter<T> {
+  protected static final int ROWS_DIVISOR = 1000;
+
+  private final List<DataFile> completedFiles = Lists.newArrayList();
+  private final PartitionSpec spec;
+  private final FileFormat format;
+  private final FileAppenderFactory<T> appenderFactory;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+
+  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+    this.spec = spec;
+    this.format = format;
+    this.appenderFactory = appenderFactory;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+  }
+
+  @Override
+  public void abort() throws IOException {
+    close();
+
+    // clean up files created by this writer
+    Tasks.foreach(completedFiles)
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  @Override
+  public List<DataFile> pollCompleteFiles() {
+    if (completedFiles.size() > 0) {
+      List<DataFile> dataFiles = ImmutableList.copyOf(completedFiles);
+      completedFiles.clear();
+      return dataFiles;
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  protected OutputFileFactory outputFileFactory() {
+    return this.fileFactory;
+  }
+
+  WrappedFileAppender createWrappedFileAppender(PartitionKey partitionKey,
+                                                Supplier<EncryptedOutputFile> outputFileSupplier) {
+    EncryptedOutputFile outputFile = outputFileSupplier.get();
+    FileAppender<T> appender = appenderFactory.newAppender(outputFile.encryptingOutputFile(), format);
+    return new WrappedFileAppender(partitionKey, outputFile, appender);
+  }
+
+  class WrappedFileAppender {
+    private final PartitionKey partitionKey;
+    private final EncryptedOutputFile encryptedOutputFile;
+    private final FileAppender<T> appender;
+
+    private boolean closed = false;
+    private long currentRows = 0;
+
+    WrappedFileAppender(PartitionKey partitionKey, EncryptedOutputFile encryptedOutputFile, FileAppender<T> appender) {
+      this.partitionKey = partitionKey;
+      this.encryptedOutputFile = encryptedOutputFile;
+      this.appender = appender;
+    }
+
+    void add(T record) {
+      this.appender.add(record);
+      this.currentRows++;
+    }
+
+    boolean shouldRollToNewFile() {
+      //TODO: ORC file now not support target file size before closed
```

Comment:
We should consider changing the ORC appender to simply return 0 if the file isn't finished. That way this check is still valid, but the file will never be rolled.

Reply:
It could be a separate issue to address this ORC issue you described ?   I think we could focus on the writer refactor.